### PR TITLE
Replace all the Decoder.if_* functions by a single Decoder.alt

### PR DESCRIPTION
**Pull request by diml**
This is part of a larger work to make `Decoder.t` a pure applicative (or selective, will see how it goes). Once we remove `Decoder.(>>=)`, we need a lot more specialised `if_xxx` functions. To avoid their proliferation, this PR proposes to replace them all by a single and general `alt` combinator that tries a list of parsers in sequence and uses the first one that succeeds:

```ocaml
val alt : 'a t list -> 'a t
```

The difficult part is to decide which error to reraise when they all fail. Ideally, we want to reraise the one from the branch that matched the most structural elements. This is difficult to implement right now, so I used a rubbish implementation that will lead to bad error messages in some cases. However, once `Decoder.t` is a pure applicative, it should be easier to change it and even show a grammar of what is accepted to the user in case of error.

### CODE REVIEW

**Comment by snowleopard** [comment]
@diml The combinator `alt` seems fine to me, although it looks like you're mostly selecting from just two alternatives, so an operator may be more convenient. In Haskell, the standard operator for selecting from two alternative parsers is `<|>`, and what you call `alt` is called `asum` (I don't like this name).

As a crazy alternative, you could add something like ``( and| ) : `a t -> `a t -> (`a * unit) t`` making it possible to write:

```ocaml
let+ res = parser1
and| () = parser2
in ...
```

:-)

**Comment by diml** [comment]
`<|>` works pretty well indeed. I replaced `alt` by this operator. The `and|` syntax looks a bit too cryptic to me :)

**Comment by snowleopard** [comment]
@diml Great! I like the version with `<|>` too.

**Comment by diml** [comment]
Just realised there is an issue for things such as `sum l <|> string`. Indeed, for constructor without arguments we accept both `(<atom>)` and `<atom>`. So `sum l <|> string` is not the same as `if_list ~then_:(sum l) ~else:string` if `l` has constant constructors. I'm adding a parameter to `sum` to prevent it from accepting the non-parenthesised in such cases.

**Comment by diml** [comment]
Alright, this is done.

**Comment by snowleopard** [comment]
It seems to me that in such situations we really want to have a selective parser, so that we can choose the subsequent parser depending on the first character, instead of trying one parser after another.

Is it difficult to implement ``select : (`a, `b) either t -> (`a -> `b) t -> `b t``?

**Comment by diml** [comment]
It shouldn't be difficult. But IMO whether we should do it depends on whether we want it for performance reasons or for clarity.

If it is only for performance reasons, then there is another option: once `Decoder.t` is an applicative we can infer such a check ourselves by seeing that one branch only accepts list and the other doesn't accepts list.

So the question is, which mental model is clearer? With what is currently in the PR, the mental model is: "the syntax is this or this". If we introduce the selective constructs and use them, I believe it would become: "if it is a list, the syntax is this otherwise it is this".

One of my hope is that we can eventually generalise `Decoder.t` so that it accepts other forms of structured input. i.e. you write one `Decoder.t` and then you can use it to parse s-expressions, canonical s-sexpressions, json, ... With this goal in mind, I feel like the first mental model is more generic and would be better suited. However, I don't want to sacrifice clarity to reach this goal. Plus maybe it's not even possible because we rely too much of the loose format of S-expressions.

So I'm curious what you and other thinks about this. In particular, which code do you find easier to reason about?

this:

```ocaml
sum ~force_parens:true l <|> decode
```

or:

```ocaml
ifS is_list ~then_:(sum l) ~else_:decode
```

**Comment by snowleopard** [comment]
I agree: the form with alternatives is clearer, it mirrors the widely-accepted BNF grammar notation. If there are no performance issues, we should probably stick to it.

The form with selective operators is indeed more about an efficient implementation: it forces one to find tokens that can distinguish branches so that appropriate parsers can be selected efficiently.

**Comment by diml** [comment]
Mirroring the BNF grammar notation is a good point. Right now the new code is less efficient, however it is only temporary until `Decoder.t` is a full applicative. And in any case, I don't expect this to be a bottleneck.

Just to make sure, I benchmarked this PR against the platform repository and observed no performance degradation. I did found a bug in the PR though...

So yeah, let's stick to it.

File: src/dune_lang/decoder.ml [hunk 1]
```diff
@@ -286,19 +281,33 @@ let enter t =
...
-    enter
-      ( junk >>= fun () ->
-        then_ >>| fun f -> f (loc, name) )
-  | _ -> else_
+let ( <|> ) =
+  (* Before you read this code, close your eyes and internalise the fact that
+     this code is temporary. It is a temporary state as part of a larger work to
+     turn [Decoder.t] into a pure applicative. Once this is done, this function
+     will be implemented in a better way and with a much cleaner semantic. *)
+  let approximate_how_much_input_a_failing_branch_consumed
```
Comment (snowleopard) [inline src/dune_lang/decoder.ml:289]: Just a note: I think parsers often disallow alternatives with overlapping prefixes to make sure there is no deep backtracking. If you do this too, then a failing parser can consume at most one symbol from the input.

Comment (diml) [reply src/dune_lang/decoder.ml:289]: That seems like a good model indeed. Thanks for mentioning this! I'll do it once the transition is over.

**Pull request merged**